File: backend/java/src/Task.java
package src;

import java.util.ArrayList;
import java.util.List;

public class Task {
    public record File(
        int id,
        String name,
        List<String> categories,
        int parent,
        int size
    ) {}

    /**
     * Task 1
     */
    public static List<String> leafFiles(List<File> files) {
        return new ArrayList<>();
    }

    /**
     * Task 2
     */
    public static List<String> kLargestCategories(List<File> files, int k) {
        return new ArrayList<>();
    }

    /**
     * Task 3
     */
    public static int largestFileSize(List<File> files) {
        return 0;
    }

    public static void main(String[] args) {
        List<File> testFiles = List.of(
            new File(1, "Document.txt", List.of("Documents"), 3, 1024),
            new File(2, "Image.jpg", List.of("Media", "Photos"), 34, 2048),
            new File(3, "Folder", List.of("Folder"), -1, 0),
            new File(5, "Spreadsheet.xlsx", List.of("Documents", "Excel"), 3, 4096),
            new File(8, "Backup.zip", List.of("Backup"), 233, 8192),
            new File(13, "Presentation.pptx", List.of("Documents", "Presentation"), 3, 3072),
            new File(21, "Video.mp4", List.of("Media", "Videos"), 34, 6144),
            new File(34, "Folder2", List.of("Folder"), 3, 0),
            new File(55, "Code.py", List.of("Programming"), -1, 1536),
            new File(89, "Audio.mp3", List.of("Media", "Audio"), 34, 2560),
            new File(144, "Spreadsheet2.xlsx", List.of("Documents", "Excel"), 3, 2048),
            new File(233, "Folder3", List.of("Folder"), -1, 4096)
        );
        
        List<String> leafFiles = leafFiles(testFiles);
        leafFiles.sort(null);
        assert leafFiles.equals(List.of(
            "Audio.mp3",
            "Backup.zip",
            "Code.py",
            "Document.txt",
            "Image.jpg",
            "Presentation.pptx",
            "Spreadsheet.xlsx",
            "Spreadsheet2.xlsx",
            "Video.mp4"
        ));

        assert kLargestCategories(testFiles, 3).equals(List.of(
            "Documents", "Folder", "Media"
        ));

        assert largestFileSize(testFiles) == 20992;
    }
}

File: backend/py/task.py
from dataclasses import dataclass
from collections import defaultdict

@dataclass
class File:
    id: int
    name: str
    categories: list[str]
    parent: int
    size: int


"""
Task 1
"""
def leafFiles(files: list[File]) -> list[str]:
    namelist = list()
    for file in files:
        isParentOf = filter(lambda x: x.parent == file.id, files)
        parentOf = list(isParentOf)
        if len(parentOf) == 0:
            namelist.append(file.name)
    return namelist


"""
Task 2
"""

def kLargestCategories(files: list[File], k: int) -> list[str]:
    categoryFrequency = {}
    for file in files:
        for cat in file.categories:
            if not cat in categoryFrequency:
                categoryFrequency[cat] = 0
            categoryFrequency[cat] += 1
    categoryFrequency = sorted(categoryFrequency.items(), key=lambda x: (-x[1],x[0]))
    kCat = list(dict(categoryFrequency).keys())[:k]
    return kCat


"""
Task 3
"""
def largestFileSize(files: list[File]) -> int:
    fileSizeDict = {}
    for file in files:
        fileSizeDict[file.name] = 0
    for file in files:
        currentFile = file
        fileSizeDict[file.name] += file.size
        while not currentFile.parent == -1:
            child = currentFile
            currentFile = list(filter(lambda x: x.id == currentFile.parent, files))[0]
            fileSizeDict[currentFile.name] += file.size
    largestSize = 0
    for key in fileSizeDict:
        if fileSizeDict[key] > largestSize:
            largestSize = fileSizeDict[key]
    return largestSize

if __name__ == '__main__':
    testFiles = [
        File(1, "Document.txt", ["Documents"], 3, 1024),
        File(2, "Image.jpg", ["Media", "Photos"], 34, 2048),
        File(3, "Folder", ["Folder"], -1, 0),
        File(5, "Spreadsheet.xlsx", ["Documents", "Excel"], 3, 4096),
        File(8, "Backup.zip", ["Backup"], 233, 8192),
        File(13, "Presentation.pptx", ["Documents", "Presentation"], 3, 3072),
        File(21, "Video.mp4", ["Media", "Videos"], 34, 6144),
        File(34, "Folder2", ["Folder"], 3, 0),
        File(55, "Code.py", ["Programming"], -1, 1536),
        File(89, "Audio.mp3", ["Media", "Audio"], 34, 2560),
        File(144, "Spreadsheet2.xlsx", ["Documents", "Excel"], 3, 2048),
        File(233, "Folder3", ["Folder"], -1, 4096),
    ]

    assert sorted(leafFiles(testFiles)) == [
        "Audio.mp3",
        "Backup.zip",
        "Code.py",
        "Document.txt",
        "Image.jpg",
        "Presentation.pptx",
        "Spreadsheet.xlsx",
        "Spreadsheet2.xlsx",
        "Video.mp4"
    ]

    

    assert kLargestCategories(testFiles, 3) == [
        "Documents", "Folder", "Media"
    ]

    assert largestFileSize(testFiles) == 20992
